src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -621,59 +628,73 @@ protected Request buildFetchRequest(URL endPoint, String accessToken) {
+   *
    * @param req request
    * @return Response
    * @throws AribaException       AribaException
    * @throws InterruptedException InterruptedException
    * @throws IOException          IOException
    */
   public Response executeRequest(Request req) throws AribaException, InterruptedException, IOException {
-    OkHttpClient enhancedOkHttpClient = getConfiguredClient().build();
-    Response response = null;
+    int actualMaxRetryCount = retryRequired ? maxRetryCount : 0;
+    RetryPolicy<Object> retryPolicy = RetryPolicy.builder()
+      .handle(AribaRetryableException.class)
+      .withBackoff(initialRetryDuration, maxRetryDuration, ChronoUnit.SECONDS, retryMultiplier)
+      .withMaxRetries(actualMaxRetryCount)
+      .onRetry(event -> LOG.info("Retrying Ariba call with plugin. Retry count: " + event.getAttemptCount()))
+      .onSuccess(event -> LOG.debug("Ariba plugin call has been executed successfully."))
+      .onRetriesExceeded(event -> LOG.error("Retry limit for the Ariba plugin has been exceeded.",
+        event.getException()))
+      .build();
+
     try {
-      response = enhancedOkHttpClient.newCall(req).execute();
-    if (response.code() != HttpURLConnection.HTTP_OK && AribaUtil.isNullOrEmpty(response.message())) {
-      AribaResponseContainer responseContainer = aribaResponse(response);
-      InputStream responseStream = responseContainer.getResponseBody();
-      JsonNode jsonNode = objectMapper.readTree(responseStream);
-      String errMsg = jsonNode.get(ResourceConstants.MESSAGE).asText() != null
-        ? jsonNode.get(ResourceConstants.MESSAGE).asText() :
-        ResourceConstants.ERR_NOT_FOUND.getMsgForKey();
-      throw new AribaException(errMsg, response.code());
-    }
-    if (!isApiLimitExhausted(response)) {
-      checkAndThrowException(response);
-      return response;
-    } else {
-      response = enhancedOkHttpClient.newCall(req).execute();
+      return Failsafe.with(retryPolicy).get(() -> executeRetryableRequest(req, actualMaxRetryCount > 0));
+    } catch (FailsafeException fse) {
+      Throwable t = fse.getCause();
+      if (t instanceof AribaException) {
+        throw (AribaException) t;
+      } else if (t instanceof InterruptedException) {
+        throw (InterruptedException) t;
+      } else if (t instanceof IOException) {
+        throw (IOException) t;
+      } else {
+        throw new RuntimeException(t);
+      }
     }
-  } catch (IOException e) {
-    throw new IOException("Endpoint is incorrect. Unable to validate the source with the provided.", e);
   }
 
+  /**
+   * Calls given Ariba API.
+   * @param req request
+   * @return Response
+   * @throws AribaException          AribaException
+   * @throws InterruptedException    InterruptedException
+   * @throws IOException             IOException
+   * @throws AribaRetryableException
+   */
+  public Response executeRetryableRequest(Request req, boolean shouldWait)
```

Comment:
How is `shouldWait` used here?

Can we please include the explanation in java doc?

Reply:
`shouldWait` is passed to `checkAndThrowException` i have added this in the doc.

> @param shouldWait Do we to wait for time defined in header if API limit is exhausted